cockroach-metrics/src/lib.rs

```diff
@@ -0,0 +1,1072 @@
+                    buckets[i - 1].le <= buckets[i].le,
+                    "Bucket {} (le={}) should be <= bucket {} (le={})",
+                    i - 1,
+                    buckets[i - 1].le,
+                    i,
+                    buckets[i].le
+                );
+            }
+        } else {
+            panic!("Expected histogram for sql_exec_latency");
+        }
+    }
+
+    #[tokio::test]
+    async fn test_cluster_client_caching() {
+        let log = slog::Logger::root(slog::Discard, slog::o!());
+        let cluster = CockroachClusterAdminClient::new(log);
+
+        // Initially no cached clients
+        assert_eq!(cluster.get_cached_addresses().await.len(), 0);
+
+        // Fetch should fail with no backends configured
+        assert!(cluster.fetch_prometheus_metrics().await.is_err());
+        assert!(cluster.fetch_node_status().await.is_err());
+
+        // Add some backends
+        let addr1: SocketAddr = "127.0.0.1:8080".parse().unwrap();
+        let addr2: SocketAddr = "127.0.0.1:8081".parse().unwrap();
+        let addr3: SocketAddr = "127.0.0.1:8082".parse().unwrap();
+
+        cluster.update_backends(&[addr1, addr2]).await;
+        let cached = cluster.get_cached_addresses().await;
+        assert_eq!(cached.len(), 2);
+        assert!(cached.contains(&addr1));
+        assert!(cached.contains(&addr2));
+
+        // Update with different set - should keep addr2, drop addr1, add addr3
+        cluster.update_backends(&[addr2, addr3]).await;
+        let cached = cluster.get_cached_addresses().await;
+        assert_eq!(cached.len(), 2);
+        assert!(!cached.contains(&addr1));
+        assert!(cached.contains(&addr2));
+        assert!(cached.contains(&addr3));
+
+        // Clear all backends
+        cluster.update_backends(&[]).await;
+        assert_eq!(cluster.get_cached_addresses().await.len(), 0);
+
+        // Fetch should fail again with no backends configured
+        assert!(cluster.fetch_prometheus_metrics().await.is_err());
+        assert!(cluster.fetch_node_status().await.is_err());
+    }
+
+    mod proptest_tests {
+        use super::*;
+        use proptest::prelude::*;
+
+        proptest! {
+            /// Test that PrometheusMetrics::parse never panics on arbitrary input
+            #[test]
+            fn prometheus_parse_never_panics(input in ".*") {
+                // This should never panic, even on completely malformed input
+                let _result = PrometheusMetrics::parse(&input);
+            }
+
+            /// Test PrometheusMetrics::parse with more structured but still arbitrary input
+            #[test]
+            fn prometheus_parse_structured_input(
+                lines in prop::collection::vec(".*", 0..20)
+            ) {
+                let input = lines.join("\n");
+                let _result = PrometheusMetrics::parse(&input);
+            }
+        }
+
+        #[test]
+        fn test_prometheus_metrics_parse_resilience() {
```

Comment:
AFAICT this isn't a proptest, so it probably shouldn't be in the `proptest_tests` module?